src/perf/index.ts

```diff
@@ -0,0 +1,133 @@
+  private readonly components: PerfComponents
+  private started: boolean
+  private readonly databuf: ArrayBuffer
+
+  constructor (components: PerfComponents) {
+    this.components = components
+    this.started = false
+    this.protocol = PROTOCOL
+    this.databuf = new ArrayBuffer(Number(writeBlockSize))
+  }
+
+  async start () {
+    await this.components.registrar.handle(this.protocol, (data: IncomingStreamData) => { void this.handleMessage(data) }, {
+      maxInboundStreams: maxStreams,
+      maxOutboundStreams: maxStreams
+    })
+    this.started = true
+  }
+
+  async stop () {
+    await this.components.registrar.unhandle(this.protocol)
+    this.started = false
+  }
+
+  isStarted () {
+    return this.started
+  }
+
+  async handleMessage (data: IncomingStreamData) {
+    const { stream } = data
+
+    let bytesToSendBack: bigint | null = null
+    for await (const buf of stream.source) {
+      if (bytesToSendBack === null) {
+        bytesToSendBack = BigInt(buf.getBigUint64(0, false))
+      }
+      // Ingest all the bufs and wait for the read side to close
+    }
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    if (bytesToSendBack === null) {
+      throw new Error('bytesToSendBack was null')
+    }
+    await stream.sink(async function * () {
+      while (bytesToSendBack > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > bytesToSendBack) {
+          toSend = bytesToSendBack
+        }
+        bytesToSendBack = bytesToSendBack - toSend
+        yield uint8Buf.slice(0, Number(toSend))
+      }
+    }())
+  }
+
+  async startPerfOnStream (peer: PeerId, sendBytes: bigint, recvBytes: bigint, options: AbortOptions = {}): Promise<void> {
+    log('dialing %s to %p', this.protocol, peer)
+
+    const uint8Buf = new Uint8Array(this.databuf)
+
+    const connection = await this.components.connectionManager.openConnection(peer, options)
+    const signal = options.signal
+    const stream = await connection.newStream([this.protocol], {
+      signal
+    })
+
+    // Convert sendBytes to uint64 big endian buffer
+    const view = new DataView(this.databuf)
+    view.setBigInt64(0, recvBytes, false)
+
+    await stream.sink((async function * () {
+      // Send the number of bytes to receive
+      yield uint8Buf.slice(0, 8)
+      // Send the number of bytes to send
+      while (sendBytes > 0n) {
+        let toSend: bigint = writeBlockSize
+        if (toSend > sendBytes) {
+          toSend = sendBytes
+        }
+        sendBytes = sendBytes - toSend
+        yield uint8Buf.slice(0, Number(toSend))
+      }
+    })())
+
+    // Read the received bytes
+    let actualRecvdBytes = BigInt(0)
+    for await (const buf of stream.source) {
+      actualRecvdBytes += BigInt(buf.length)
+    }
+
+    if (actualRecvdBytes !== recvBytes) {
+      throw new Error(`Expected to receive ${recvBytes} bytes, but received ${actualRecvdBytes}`)
+    }
+
+    stream.close()
```

Comment:
Shouldn't the node close the write side of the stream right after having sent the last byte of its payload?

```
4. Close the write side of our stream.
5. Read from the read side of the stream. This  should be the same number of bytes as we told the server in step 2.
```

Or is write-side closing happening implicitly through `stream.sink` above?